### PR TITLE
feat: introduce `:outdated` pseudo selector

### PR DESCRIPTION
**Pull request by nlf**
this adds the `:outdated` pseudo selector, along with documentation

### CODE REVIEW

File: workspaces/arborist/lib/query-selector-all.js [hunk 1]
```diff
@@ -422,6 +431,135 @@ class Results {
...
+        return false
+      }
+
+      // we cache the promise representing the full versions list, this helps reduce the
+      // number of requests we send by keeping population of the cache in a single tick
+      // making it less likely that multiple requests for the same package will be inflight
+      if (!this.#outdatedCache.has(node.name)) {
+        this.#outdatedCache.set(node.name, getPackageVersions(node.name, this.flatOptions))
+      }
+      const availableVersions = await this.#outdatedCache.get(node.name)
```
Comment (wraithgar) [inline workspaces/arborist/lib/query-selector-all.js:454]: nit: awaiting is "expensive" and if we `set` the awaited value we are only doing it once.

File: docs/content/using-npm/dependency-selectors.md [hunk 2]
```diff
@@ -78,6 +79,28 @@ Some examples:
...
+- `out-of-range` a version exists that is greater than the current one, does not satisfy at least one of its dependents
+- `major` a version exists that is a semver major greater than the current one
+- `minor` a version exists that is a semver minor greater than the current one
+- `patch` a version exists that is a semver patch greater than the current one
+
+In addition to the filtering performed by the pseudo selector, some extra data is added to the resulting objects. The following data can be found under the `queryContext` property of each node.
+
+- `versions` an array of every available version of the given node
+- `outdated.inRange` an array of objects, each with a `from` and `versions`, where `from` is the on-disk location of the node that depends on the current node and `versions` is an array of all available versions that satisfies that dependency. this is only populated if `:outdated(in-range)` is used
+- `outdated.outOfRange` an array of objects, identical in shape to `inRange`, but where the `versions` array is every available version that does not satisfy the dependency
```
Comment (wraithgar) [inline docs/content/using-npm/dependency-selectors.md]:
```suggestion
- `outdated.inRange` an array of objects, each with a `from` and `versions`, where `from` is the on-disk location of the node that depends on the current node and `versions` is an array of all available versions that satisfies that dependency. This is only populated if `:outdated(in-range)` is used.
- `outdated.outOfRange` an array of objects, identical in shape to `inRange`, but where the `versions` array is every available version that does not satisfy the dependency. This is only populated if `:outdated(out-of-range)` is used.
```

**Review by wraithgar — CHANGES_REQUESTED** [review]
Docs update and a take-it-or-leave-it nit.

File: workspaces/arborist/lib/query-selector-all.js (hunk 1 above)
Comment (nlf) [reply workspaces/arborist/lib/query-selector-all.js:454]: in my admittedly limited testing, the overhead of awaiting this value repeatedly was vastly smaller than the repeated work done by awaiting a result before storing it in the cache (~4.5 seconds awaiting before storing, ~3.1 seconds storing before awaiting on my machine in this repo)

that being the case, i think i'd like to leave this as-is for now and we can always tweak it later

Comment (wraithgar) [reply workspaces/arborist/lib/query-selector-all.js:454]: boom. resolved.

File: docs/content/using-npm/dependency-selectors.md (hunk 2 above)
Comment (nlf) [reply docs/content/using-npm/dependency-selectors.md]: :+1: done

**Review by wraithgar — APPROVED** [review]

**Pull request merged**